src/sycl/GroupGemm.cpp

```diff
@@ -0,0 +1,223 @@
+#include "kernels/moe/xe_array_mma.hpp"
+#include "kernels/moe/xe_moe_gemm.hpp"
+
+using namespace cute;
+
+template <typename scalar_t>
+struct MoERunner {
+  using ProblemShape = cutlass::gemm::GroupProblemShape<Shape<int, int, int>>;  // <M,N,K> per group
+  template <typename Gemm>
+  typename Gemm::Arguments args_from_options(
+      const cutlass::KernelHardwareInfo& hw_info,
+      const typename Gemm::ElementA* A_ptr,
+      const typename Gemm::ElementB* B_ptr,
+      typename Gemm::CollectiveEpilogue::ElementOutput* D_ptr,
+      const int gemm_N,
+      const int gemm_K,
+      const int* num_rows_per_expert_device,
+      const int num_experts) {
+    typename Gemm::Arguments arguments;
+    decltype(arguments.fusion_args) fusion_args;
+
+    fusion_args.alpha = 1;
+    fusion_args.beta = 0;
+    fusion_args.alpha_ptr = nullptr;
+    fusion_args.beta_ptr = nullptr;
+    fusion_args.alpha_ptr_array = nullptr;
+    fusion_args.beta_ptr_array = nullptr;
+    // One alpha and beta per each group
+    fusion_args.dAlpha = {cute::_0{}, cute::_0{}, 1};
+    fusion_args.dBeta = {cute::_0{}, cute::_0{}, 1};
+
+    using RasterOrderOptions =
+        typename cutlass::gemm::kernel::detail::PersistentTileSchedulerXeGroup<ProblemShape>::RasterOrderOptions;
+
+    arguments = typename Gemm::Arguments{
+        cutlass::gemm::GemmUniversalMode::kGrouped,
+        static_cast<const typename Gemm::ElementA**>((void*)A_ptr),
+        static_cast<const typename Gemm::ElementB**>((void*)B_ptr),
+        nullptr,  // static_cast<const ElementC**>((void*)D_ptr),
+        static_cast<typename Gemm::CollectiveEpilogue::ElementOutput**>((void*)D_ptr),
+        fusion_args,
+        hw_info,
+        {1, RasterOrderOptions::AlongN},
+        num_rows_per_expert_device,
+        num_experts,
+        gemm_N,
+        gemm_K};
+
+    return arguments;
+  }
+
+  void
+  run(sycl::queue queue,
+      const void* activations,
+      const void* weights,
+      void* outputs,
+      const int gemm_n,
+      const int gemm_k,
+      const int* num_rows_per_expert_device,
+      const int num_experts) {
+    // The KernelHardwareInfo struct holds the number of EUs on the GPU with a
+    // given device ID. This information is used by the underlying kernel.
+    cutlass::KernelHardwareInfo hw_info;
+    // Change device_id to another value if you are running on a machine with
+    // multiple GPUs and wish to use a GPU other than that with device ID 0.
+    hw_info.sm_count = cutlass::KernelHardwareInfo::query_device_multiprocessor_count(hw_info.device_id);
+
+    using LayoutA = cutlass::layout::RowMajor;
+    using LayoutB = cutlass::layout::ColumnMajor;
+    using LayoutC = cutlass::layout::RowMajor;
+    using LayoutD = cutlass::layout::RowMajor;
+
+    using GmemTiledCopyA = XE_2D_U16x8x32_LD_N;
+    using GmemTiledCopyB = XE_2D_U16x16x16_LD_T;
+
+    // Workgroup-level tile
+    using TileShape = Shape<_256, _256, _32>;
```

Comment:
Please consider using a smaller value of Workgroup tile M dimension for decoding. Maybe 8 or 16.